client/asset/dcr/dcr.go

```diff
@@ -3291,50 +3171,54 @@ func (dcr *ExchangeWallet) getBestBlock(ctx context.Context) (*block, error) {
 func (dcr *ExchangeWallet) mainchainAncestor(ctx context.Context, blockHash *chainhash.Hash) (*chainhash.Hash, int64, error) {
 	checkHash := blockHash
 	for {
-		checkBlock, err := dcr.wallet.GetBlockHeaderVerbose(ctx, checkHash)
+		checkBlock, isMainchain, _, err := dcr.blockHeader(ctx, checkHash)
 		if err != nil {
 			return nil, 0, fmt.Errorf("getblockheader error for block %s: %w", checkHash, err)
 		}
-		if checkBlock.Confirmations > -1 {
+		if isMainchain {
 			// This is a mainchain block, return the hash and height.
 			return checkHash, int64(checkBlock.Height), nil
 		}
 		if checkBlock.Height == 0 {
-			return nil, 0, fmt.Errorf("no mainchain ancestor for block %s", blockHash.String())
-		}
-		checkHash, err = chainhash.NewHashFromStr(checkBlock.PreviousHash)
-		if err != nil {
-			return nil, 0, fmt.Errorf("error decoding previous hash %s for block %s: %w",
-				checkBlock.PreviousHash, checkHash.String(), err)
+			return nil, 0, fmt.Errorf("no mainchain ancestor for block %s", blockHash)
 		}
+		checkHash = &checkBlock.PrevBlock
 	}
 }
 
-func (dcr *ExchangeWallet) isMainchainBlock(ctx context.Context, block *block) (bool, error) {
-	blockHeader, err := dcr.wallet.GetBlockHeaderVerbose(ctx, block.hash)
+// blockHeader returns the *BlockHeader for the specified block hash, and a
+// validity bool indicating true when the block has not been stake-invalidated.
+func (dcr *ExchangeWallet) blockHeader(ctx context.Context, blockHash *chainhash.Hash) (blockHeader *BlockHeader, valid, mainchain bool, err error) {
```

Comment:
I could also add `Valid, Mainchain bool` fields to the `BlockHeader` and let the implementation decide how to fill them in.

Reply:
Hmm, yeah that might work out nicely.  Might also be worth documenting that it will never set valid = true where mainchain = false (if I'm reading the code right).